### PR TITLE
feat(storage): noncurrent and custom time OLM

### LINKED ISSUE #4275: Support daysSinceNoncurrentTime and noncurrentTimeBefore OLM options
Basically this change entails

- [x] Adding two fields to `storage::Lifecycle::Condition`:

```C++
struct storage::LifecycleRule::Condition {
 ....
  optional<std::int32_t> days_since_noncurrent_time;
  optional<std::chrono::system_clock::time_point> noncurrent_time_before;
};
```

- [x] Adding helper functions to create conditions with one of these fields set
- [x] Change the existing code to merge these fields as part of `*Conjunction`

### LINKED ISSUE #4432: Storage: support customTime metadata and custom timestamp based OLM
Request to support a new customTime metadata and associated custom timestamp based OLM conditions:

```cpp
std::chrono::time_point tenDays = std::chrono::system_clock::now() + 
                                      std::chrono::days{10};

gcs::BucketLifecycle bucket_lifecycle_rules = gcs::BucketLifecycle{
    {gcs::LifecycleRule(
         gcs::LifecycleRule::ConditionConjunction(
             gcs::LifecycleRule::DaysSinceCustomTime(30),
             gcs::LifecycleRule::CustomTimeBefore(tenDays)),
          gcs::LifecycleRule::Delete())}};

StatusOr<gcs::BucketMetadata> updated_metadata = client.PatchBucket(
    bucket_name,
      gcs::BucketMetadataPatchBuilder().SetLifecycle(bucket_lifecycle_rules));

// New metadata field:
StatusOr<gcs::ObjectMetadata> object_metadata =
    client.GetObjectMetadata(bucket_name, object_name);

gcs::ObjectMetadata desired = *object_metadata;
desired.mutable_metadata().emplace("customTime", "2020-01-01T00:00:00Z");

StatusOr<gcs::ObjectMetadata> updated =
    client.UpdateObject(bucket_name, object_name, desired,
                        gcs::Generation(object_metadata->generation()));
```

### PR DESCRIPTION
**Pull request by coryan**
Implement new Object Lifecycle Management (OLM) conditions
based on:

- "noncurrent time", i.e., the time since the object version is
"archived" or no longer the "current" version.
- "custom time", i.e., a custom field in the object metadata.

This fixes #4275 and fixes #4432

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4871)

### CODE REVIEW

File: google/cloud/storage/internal/bucket_requests.cc [hunk 1]
```diff
@@ -85,8 +85,9 @@ StatusOr<LifecycleRule> LifecycleRuleParser::FromJson(
       auto const date = condition.value("createdBefore", "");
       absl::CivilDay day;
       if (!absl::ParseCivilTime(date, &day)) {
-        return Status(StatusCode::kInvalidArgument,
-                      "Cannot parse " + date + " as a date");
+        return Status(
+            StatusCode::kInvalidArgument,
+            "Cannot parse createdBefore value (" + date + "( as a date");
```
Comment (devbww) [inline google/cloud/storage/internal/bucket_requests.cc:90]: s/( as a/) as a/

File: google/cloud/storage/lifecycle_rule.cc [hunk 2]
```diff
@@ -148,6 +166,38 @@ void LifecycleRule::MergeConditions(LifecycleRuleCondition& result,
       result.num_newer_versions.emplace(std::forward<std::int32_t>(tmp));
     }
   }
+  if (rhs.days_since_noncurrent_time.has_value()) {
+    if (result.days_since_noncurrent_time.has_value()) {
+      *result.days_since_noncurrent_time = (std::max)(
```
Comment (devbww) [inline google/cloud/storage/lifecycle_rule.cc:171]: Q: Do we need an indirection here when we don't 3 lines down?

Comment (coryan) [reply google/cloud/storage/lifecycle_rule.cc]: No, fixed here and everywhere else in this function.

File: google/cloud/storage/internal/bucket_requests.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/bucket_requests.cc]: Fixed.

**Comment by coryan** [comment]
PTAL

**Pull request merged**